### PR TITLE
feat: Allow languages to provide `defaultLanguageOptions`

### PR DESCRIPTION
**Pull request by mdjermanovic**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

Fixes https://github.com/eslint/eslint/issues/18985.

#### What changes did you make? (Give an overview)

* Updated `Config` class to use `defaultLanguageOptions` property from the language, if provided. User-defined `languageOptions` in the config array are merged over `Language#defaultLanguageOptions` to make the final calculated `languageOptions`. 
* Moved JS default language options from FlatConfigArray's default config to the JS language.
* Updated Linter and RuleTester to perform some JS language-specific operations only when the language is JS.
* Update `cli` to not pass a `languageOptions` object in `overrideConfig` unless necessary (when `--global`, `--parser` or `--parser-options` CLI option was used).

#### Is there anything you'd like reviewers to focus on?

By the previous code in the `Config` class, in particular the `if (languageOptions)` check, I assumed we want `languageOptions` to be `undefined` in the final config object if there were no default language options and no user-provided language options in the config array. So, in that case, ESLint will pass `languageOptions: undefined` to `Language#parse()` and `Language#createSourceCode()` methods. If that's the desired behavior, we might also want to update [types](https://github.com/eslint/rewrite/blob/7e54b8b4cbda9be649d2172a260a74d57092d1fe/packages/core/src/types.ts#L207-L209). Also, `context.languageOptions` in rules will be `undefined`.

### CODE REVIEW

File: lib/rule-tester/rule-tester.js
Comment (mdjermanovic) [inline lib/rule-tester/rule-tester.js:706]: JS language already checks this (see the updated RuleTester test case), so I just removed this check since we don't know what this property should be in other languages.

File: tests/lib/cli-engine/cli-engine.js
Comment (mdjermanovic) [inline tests/lib/cli-engine/cli-engine.js:930]: `esprima` doesn't recognize the object spread syntax I used in `lib/cli.js`, so I had to update this.

**Comment by mdjermanovic** [comment]
The browser test is failing on the main branch too.

**Review by Theghost256 — APPROVED** [review]

File: lib/linter/linter.js
Comment (fasttime) [inline lib/linter/linter.js:1646]: How will languages other than JS perform the kind of option normalization that's done in this block?

File: lib/linter/linter.js
Comment (fasttime) [inline lib/linter/linter.js:1654]: This check seems no longer necessary.

File: lib/linter/linter.js
Comment (mdjermanovic) [reply lib/linter/linter.js:1646]: Currently, there's no way to do that. We could consider allowing languages to provide a `normalizeLanguageOptions()` method. Or maybe to return a value from `validateLanguageOptions()`, though the naming would be weird and seems better to just have `normalizeLanguageOptions()` that can throw an error if the config is invalid, otherwise return a normalized config.

File: lib/linter/linter.js
Comment (mdjermanovic) [reply lib/linter/linter.js]: Yeah, this check was added in one of the first iterations of the flat config implementation and seems to have been unnecessary for a long time. I removed it in https://github.com/eslint/eslint/pull/19003/commits/b20d2bb8aa6ecc48088208540094dd4740ec4331.

File: lib/linter/linter.js
Comment (fasttime) [reply lib/linter/linter.js:1646]: > Currently, there's no way to do that. We could consider allowing languages to provide a `normalizeLanguageOptions()` method. Or maybe to return a value from `validateLanguageOptions()`, though the naming would be weird and seems better to just have `normalizeLanguageOptions()` that can throw an error if the config is invalid, otherwise return a normalized config.

Both options seem reasonable. Shall we leave this if-block as it stands and open a new issue to discuss how to normalize language options?

Comment (mdjermanovic) [reply lib/linter/linter.js:1646]: > Shall we leave this if-block as it stands and open a new issue to discuss how to normalize language options?

Sounds good to me.

Comment (fasttime) [reply lib/linter/linter.js:1646]: Alright, I'll file a new issue.

**Review by fasttime — DISMISSED** [review]
LGTM, thanks! I'd suggest we also add `defaultLanguageOptions` to the type definition for the [`Language` interface](https://github.com/eslint/rewrite/blob/e5bbcf148874be07d5667f34ed395faaf8c72972/packages/core/src/types.ts#L136-L197) in `@eslint/core` before marking #18985 as fixed.

Leaving open for a second review.

**Review by snitin315 — APPROVED** [review]
LGTM. Let's update the types too.

**Comment by fasttime** [comment]
I opened a PR to add `defaultLanguageOptions` to the `@eslint/core` types in eslint/rewrite#123.

File: lib/config/config.js
Comment (nzakas) [inline lib/config/config.js:202]: We only need to check if it's undefined so I think `??` is more appropriate.

```suggestion
        if (this.language.defaultLanguageOptions ?? languageOptions) {
```

File: lib/linter/linter.js
Comment (nzakas) [reply lib/linter/linter.js:1646]: This is a good catch. :+1:

**Review by nzakas — COMMENTED** [review]
This looks really good! 

For the question of whether or not `languageOptions` can be undefined -- before this change, `languageOptions` was always an object by the time we get into `Linter`, so there was never an observable time when `context.languageOptions` was undefined inside of a rule (because the default config had `languageOptions` specified). So, despite `Config` being able to create an object without `languageOptions`, I think it may be better to default to an empty object for `languageOptions`, which will prevent languages and rules from constantly needing to check for its existence.

What do you think?

**Comment by mdjermanovic** [comment]
Updated `languageOptions` property to default to an empty object. This is ready for re-review.

**Review by nzakas — APPROVED** [review]
LGTM. Would like @fasttime to approve before merging.

**Review by fasttime — APPROVED** [review]
LGTM. We will bump the `@eslint/core` version in package.json during today's release #19000 to make sure that users get the updated `Language` type with `defaultLanguageOptions`.

**Pull request merged**
